### DISCUSSION TITLE
Ability to search on multiple parameters instead of only displayName?

### DISCUSSION THREAD

**Discussion by mariaarcena** [YannickRe/spfx-msgraph-peoplesearch#42] [Ideas]
@YannickRe Hello Yannick! ✨Just got to test your people search webpart recently because it was recommended by one our support engineers at Microsoft! We love it so much since it uses Microsoft Graph! 👏

I did notice that it only searches on parameter "displayName"---I think it could be more enhanced by having the ability to search on the given parameters inside the property pane (screenshot below) or any other way to search on other desired parameters like department, jobTitle, employeeId, etc. How easy/difficult is it to achieve that? ✨

(*forgive me if it sounds like a dumb question as I am a newbie to the world of sharepoint framework this mid year and just found out about pnp microsoft 365 community a few months ago!)
![image](https://user-images.githubusercontent.com/17426994/142302227-ce0e175b-57e3-4319-9518-bdfdf24142e6.png)

**Comment by YannickRe** [comment]
Thank you for the nice words, glad you enjoy the work!

Initially, the web part was intended to show off the new `$search` ability, so that's why it only includes a search on `displayName`. Any other property included in `$search` would fall back to the `$filter` functionality of using `$startsWith`.
The question has been asked before in issue #5, it shows the necessary code changes.

Essentially, you adjust the following line:
https://github.com/YannickRe/spfx-msgraph-peoplesearch/blob/e614c8990ad3ecf47a4317ee8834b1cca5e61bea/src/services/SearchService/SearchService.ts#L60

into something like:
```
resultQuery = resultQuery.query({ $search: `"displayName:${this.searchParameter}" OR "jobTitle:${this.searchParameter}" OR "department:${this.searchParameter}" OR "officeLocation:${this.searchParameter}" OR "businessPhones:${this.searchParameter}"` });
```

You could use `this.selectParameter` to loop over all properties to dynamically build the `$search` query.

I hope this answers your question?

**Reply by MMAtHome** [reply]
Quick question, for clarity, if we use resultQuery.query with, for example, jobTitle, will the search only do an exact match rather than a "startsWith".  Is there any way of telling how an attribute will behave?

**Reply by YannickRe** [reply]
No, if the property is not supported for full text search, it will fall back to startsWith automatically.
It looks like you want to deeply understand the serverside behavior, so I recommend you read this: https://docs.microsoft.com/en-us/graph/search-query-parameter?WT.mc_id=M365-MVP-5003400

**Comment by MMAtHome** [comment]
Yannick,

Thanks so much for taking the time to pass on this information.

Much appreciated.

Martin

**Comment by MMAtHome** [comment]
Good evening.
I was so pleased to see that you have resolved a couple of issues and also
added the option to search by some additional parameters.  I have tried to
compile the new code today and am presented with the following issue.

Error - [tsc] src/services/TemplateService/TemplateService.ts(142,17):
error TS2322: Type 'number | IComponentFieldsConfiguration[]' is not
assignable to type 'any[]'.

Is there a quick fix I could apply?

Regards,

Martin

**Reply by YannickRe** [reply]
Weird, it does compile for me...
I'd try:
- run `gulp clean`
- delete the node_modules folder and run `npm install`
- run `gulp bundle --ship` and run `gulp package-solution --ship`

If it still doesn't work, I'd like the full output from the gulp bundle command.

**Reply by DavidOsborne365** [reply]
Hi @YannickRe , I also saw the same compiling error, is this the output you were looking for:

c:\My Git Projects\spfx-msgraph-peoplesearch-2.8.3> gulp build        
Build target: DEBUG
[14:01:56] Using gulpfile ~\My Git Projects\spfx-msgraph-peoplesearch-2.8.3\gulpfile.js
[14:01:56] Starting 'build'...
[14:01:56] Starting gulp
[14:01:56] Starting subtask 'pre-copy'...
[14:01:56] Finished subtask 'pre-copy' after 144 ms
[14:01:56] Starting subtask 'copy-static-assets'...
[14:01:56] Starting subtask 'sass'...
Browserslist: caniuse-lite is outdated. Please run:
  npx update-browserslist-db@latest
  Why you should do it regularly: https://github.com/browserslist/update-db#readme
[14:01:56] Finished subtask 'sass' after 221 ms
[14:01:56] Starting subtask 'lint'...
[14:01:56] [lint] eslint version: 8.7.0
[14:01:56] Starting subtask 'tsc'...
[14:01:56] [tsc] typescript version: 4.7.4
[14:01:56] Finished subtask 'copy-static-assets' after 347 ms
[14:02:01] Error - [tsc] src/services/TemplateService/TemplateService.ts(142,17): error TS2322: Type 'number | IComponentFieldsConfiguration[]' is not assignable to type 'any[]'.
[14:02:01] [tsc] Type 'number' is not assignable to type 'any[]'.
[14:02:01] Error - 'tsc' sub task errored after 4.89 s 
 exited with code 2
[14:02:01] 'build' errored after 5.34 s
About to exit with code: 1

**Reply by YannickRe** [reply]
Thank you @davidosborneafton , that's indeed what I needed.
@MMAtHome the compile error should be fixed now.

**Reply by DavidOsborne365** [reply]
> Thank you @davidosborneafton , that's indeed what I needed. @MMAtHome the compile error should be fixed now.

Happy New Year @YannickRe. I was just testing the latest 2.8.4 version and it looks like the fix for the compile error has broken the search box functionality. 
(Just want to add that I really appreciate all you do to provide this WebPart to the community).

**Reply by YannickRe** [reply]
@davidosborneafton Can you create an issue for this? And provide me more information, like what you are trying, how it is set up, what errors are you getting or behavior are you seeing...

**Comment by MMAtHome** [comment]
Yannick,

Sorry, but it looks like the same problem still exists.

Any ideas would be much appreciated.

Regards,

Martin

**Comment by MMAtHome** [comment]
Yannick,

Thank you - this now compiles just fine for me.  There is one small issue
remaining in that if the user enters ONLY a # sign in the search, this
still generates an error message.  I think this can be communicated to the
users, but thought you would want to know.

Once again, thank you for your help with this.

Very best regards,

Martin
